ingress: Create FromGroups resource

Alex-Waring opened the pull request:
Duplicates the structures inplace to evaluate the toGroups resource into the ingress section, allowing the creation of FromGroups. This means AWS SG groups can be included as ingress resources and directly translated into fromCIDR rules.

Example usage:

```
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
metadata:
  name: "l4-rule"
spec:
  endpointSelector:
    matchLabels:
      role: backend
  ingress:
  - toPorts:
    - ports:
      - port: '80'
        protocol: TCP
    fromGroups:
    - aws:
        securityGroupsIds:
        - 'sg-0123456789abcdefg'
```

Fixes: https://github.com/cilium/cilium/issues/30032

```release-note
Allows for using AWS SGs in the ingress section of rules.
```

christarazi commented on pkg/policy/api/ingress.go:
```diff
@@ -72,6 +74,19 @@ type IngressCommonRule struct {
+	//     - 'sg-XXXXXXXXXXXXX'
+	//
+	// +kubebuilder:validation:Optional
+	FromGroups []Groups `json:"FromGroups,omitempty"`
```
```suggestion
	FromGroups []Groups `json:"fromGroups,omitempty"`
```

to conform with the other fields.

christarazi commented on pkg/policy/api/ingress.go:
```diff
@@ -237,3 +252,60 @@ func (i *IngressCommonRule) GetSourceEndpointSelectorsWithRequirements(requireme
+	newRule.FromGroups = nil
+	e.SetAggregatedSelectors()
+	return newRule, nil
+}
```
I'd suggest looking to see if we can reasonably DRY up this code from its egress counterpart.

Alex-Waring replied:
Hey Chris, I'll tidy everything else up but looking at this code I'm not sure it really makes a huge amount of sense to make it more DRY. I know it's repetitive but most of the logic is calls to struct methods, so abstracting this out was just getting ridiculously complicated.

christarazi replied:
Could we at least have a common utility function that accepts the `api.Groups` type and returns the slice of CIDR sets? Basically, it could abstract away the for-loop above.

christarazi reviewed (commented):
Thanks for the PR, initial approach LGTM. A few comments below. I'd also suggest making the rename of `ToGroups` to `Groups` a separate commit, just to help scope down the review.

christarazi commented on pkg/policy/groups/helpers.go:
I think this belongs in the rename commit as well as the corresponding test file?

christarazi commented on pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml:
You could squash these generator changes with the commit that introduced the changes.

christarazi reviewed (approved):
Thanks, looking good! Could you take care of the two comments below and it's good to go from my perspective?

Alex-Waring committed:
ingress: Rename FromGroups resource

In preparation for making the groups resource applicable to both ingress and egress
rules, this commit changes the name of the ToGroups struct to Groups.

Signed-off-by: Alex Waring <ajmwaring@gmail.com>

Alex-Waring committed:
abstract out ExtractCidrSet

As I am going to be copying the Create Derivative functions, this commit abstracts out some
of the logic into a helper function, to make the code more DRY.

Signed-off-by: Alex Waring <ajmwaring@gmail.com>

Alex-Waring committed:
ingress: Create FromGroups resource

Duplicates the structures inplace to evaluate the toGroups resource into the ingress
section, allowing the creation of FromGroups. This means AWS SG groups can be
included as ingress resources and directly translated into fromCIDR rules.

Fixes: #30032
Signed-off-by: Alex Waring <ajmwaring@gmail.com>
